File: packages/java/cnos-java/src/test/java/ai/kitsy/cnos/CnosTest.java
package ai.kitsy.cnos;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Singleton / module-level API contract tests.
 *
 * Covers all 15 contract methods:
 *   ready, read, require, readOr, value, secret, meta,
 *   setDefaultRuntime, defaultRuntime, resetDefaultRuntime,
 *   toObject, toPublicEnv, format, refreshSecrets, refreshSecret
 *
 * Also covers the library composition model:
 *   root → libA → libB → libC → libD, libE → libF
 */
class CnosTest {

    private static final String MINIMAL = "{"
            + "\"version\":1,"
            + "\"workspace\":\"base\","
            + "\"profile\":\"local\","
            + "\"resolvedAt\":\"2024-01-01T00:00:00Z\","
            + "\"configHash\":\"abc123\","
            + "\"values\":{"
            + "  \"server.port\":3000,"
            + "  \"app.name\":\"cnos-java\""
            + "},"
            + "\"derived\":{"
            + "  \"app.effectiveHost\":{"
            + "    \"expr\":\"coalesce(request.headers.host,'default.host')\","
            + "    \"deps\":[],"
            + "    \"runtimeRefs\":[\"request.headers.host\"]"
            + "  }"
            + "},"
            + "\"secretRefs\":{},"
            + "\"publicKeys\":[\"app.name\"],"
            + "\"runtimeNamespaces\":[\"request\"],"
            + "\"meta\":{"
            + "  \"workspace\":\"base\","
            + "  \"profile\":\"local\","
            + "  \"cnos_version\":\"1.14.0\""
            + "}"
            + "}";

    private static CnosRuntime makeRuntime() throws CnosError {
        return CnosRuntime.load(CnosOptions.builder()
                .projectionData(MINIMAL.getBytes(StandardCharsets.UTF_8))
                .build());
    }

    @BeforeEach
    void resetBefore() {
        Cnos.resetDefaultRuntime();
    }

    @AfterEach
    void resetAfter() {
        Cnos.resetDefaultRuntime();
    }

    // ================================================================
    // Lifecycle
    // ================================================================

    @Test
    void readBeforeInitThrows() {
        CnosError err = assertThrows(CnosError.class, () -> Cnos.read("value.server.port"));
        assertTrue(err.getMessage().contains("not initialized") || err.getMessage().contains("not ready"));
    }

    @Test
    void setDefaultRuntimeMakesReadWork() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        Optional<Object> v = Cnos.value("server.port");
        assertTrue(v.isPresent());
        assertEquals(3000, ((Number) v.get()).intValue());
    }

    @Test
    void defaultRuntimeReturnsSetInstance() throws CnosError {
        CnosRuntime rt = makeRuntime();
        Cnos.setDefaultRuntime(rt);
        assertSame(rt, Cnos.defaultRuntime());
    }

    @Test
    void resetDefaultRuntimeClears() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        Cnos.resetDefaultRuntime();
        assertThrows(CnosError.class, () -> Cnos.read("value.server.port"));
    }

    @Test
    void readyIsIdempotent() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        CnosRuntime first = Cnos.defaultRuntime();
        // ready() on already-initialized singleton — may fail to find projection on disk
        try { Cnos.ready(); } catch (CnosError ignored) {}
        assertSame(first, Cnos.defaultRuntime());
    }

    // ================================================================
    // read / require / readOr
    // ================================================================

    @Test
    void readReturnsValueForExistingKey() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        Optional<Object> v = Cnos.read("value.app.name");
        assertTrue(v.isPresent());
        assertEquals("cnos-java", v.get());
    }

    @Test
    void readReturnsEmptyForMissingKey() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        Optional<Object> v = Cnos.read("value.does.not.exist");
        assertFalse(v.isPresent());
    }

    @Test
    void requireReturnsValue() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        assertEquals("cnos-java", Cnos.require("value.app.name"));
    }

    @Test
    void requireThrowsForMissingKey() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        assertThrows(CnosError.class, () -> Cnos.require("value.does.not.exist"));
    }

    @Test
    void readOrReturnsFallback() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        assertEquals("fallback", Cnos.readOr("value.missing", "fallback"));
    }

    // ================================================================
    // value / secret / meta
    // ================================================================

    @Test
    void valueShortcut() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        Optional<Object> v = Cnos.value("server.port");
        assertTrue(v.isPresent());
        assertEquals(3000, ((Number) v.get()).intValue());
    }

    @Test
    void secretReturnsMissingWhenNoSecretRefs() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        assertFalse(Cnos.secret("does.not.exist").isPresent());
    }

    @Test
    void metaReturnsWorkspace() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        Optional<Object> v = Cnos.meta("workspace");
        assertTrue(v.isPresent());
        assertEquals("base", v.get());
    }

    // ================================================================
    // toObject / toPublicEnv / format
    // ================================================================

    @Test
    void toObjectReturnsNonEmptyMap() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        Map<String, Object> obj = Cnos.toObject();
        assertNotNull(obj);
        assertFalse(obj.isEmpty());
    }

    @Test
    void toPublicEnvIncludesPromotedKeys() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        Map<String, String> env = Cnos.toPublicEnv(new ToPublicEnvOptions("vite", ""));
        assertEquals("cnos-java", env.get("VITE_APP_NAME"));
    }

    @Test
    void formatSubstitutesConfigKeys() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        assertEquals("App: cnos-java", Cnos.format("App: ${value.app.name}"));
    }

    @Test
    void formatLeavesUnknownKeysUnchanged() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        assertEquals("${value.does.not.exist}", Cnos.format("${value.does.not.exist}"));
    }

    // ================================================================
    // refreshSecrets / refreshSecret
    // ================================================================

    @Test
    void refreshSecretsCompletesWithNoSecrets() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        assertDoesNotThrow(Cnos::refreshSecrets);
    }

    @Test
    void refreshSecretOnMissingKeyIsNoop() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());
        assertDoesNotThrow(() -> Cnos.refreshSecret("does.not.exist"));
    }

    // ================================================================
    // Composition model
    // ================================================================
    //
    // Simulates: root → libA → libB → libC → libD, libE → libF
    // Libraries read from the singleton. Only the root calls setDefaultRuntime.

    private static Optional<Object> libF_readMeta() throws CnosError {
        return Cnos.meta("workspace");
    }

    private static Optional<Object> libE_readMeta() throws CnosError {
        return libF_readMeta();
    }

    private static Optional<Object> libD_readPort() throws CnosError {
        return Cnos.value("server.port");
    }

    private static Optional<Object> libC_readPort() throws CnosError {
        return libD_readPort();
    }

    private static Object[] libB_read() throws CnosError {
        return new Object[]{libC_readPort().orElse(null), libE_readMeta().orElse(null)};
    }

    private static Object[] libA_read() throws CnosError {
        return libB_read();
    }

    @Test
    void compositionLibrariesSucceedAfterRootInitializes() throws CnosError {
        Cnos.setDefaultRuntime(makeRuntime());  // root initializes once

        Object[] results = libA_read();         // libA → … → libD/libF
        assertEquals(3000, ((Number) results[0]).intValue());
        assertEquals("base", results[1]);
    }

    @Test
    void compositionLibrariesFailBeforeRootInitializes() {
        // Root has NOT initialized — all library reads should fail
        assertThrows(CnosError.class, () -> libA_read());
    }

    @Test
    void compositionMultipleLibrariesShareSameRuntime() throws CnosError {
        CnosRuntime rt = makeRuntime();
        Cnos.setDefaultRuntime(rt);

        assertSame(rt, Cnos.defaultRuntime());
        libA_read();
        assertSame(rt, Cnos.defaultRuntime());  // unchanged after reads
    }
}

File: packages/kotlin/cnos-kotlin/src/test/kotlin/ai/kitsy/cnos/CnosTest.kt
package ai.kitsy.cnos

import java.util.Optional
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class CnosTest {

    private val MINIMAL = """
    {
      "version": 1,
      "workspace": "base",
      "profile": "local",
      "resolvedAt": "2024-01-01T00:00:00Z",
      "configHash": "abc123",
      "values": {
        "server.port": 3000,
        "app.name": "cnos-kotlin"
      },
      "derived": {
        "app.effectivePort": {
          "expr": "coalesce(process.env.PORT, value.server.port, '3000')",
          "deps": ["value.server.port"],
          "runtimeRefs": ["process.env.PORT"]
        }
      },
      "secretRefs": {},
      "publicKeys": ["app.name"],
      "runtimeNamespaces": [],
      "meta": {
        "workspace": "base",
        "profile": "local",
        "cnos_version": "1.14.0"
      }
    }""".trimIndent()

    @AfterEach
    fun cleanup() {
        Cnos.resetDefaultRuntime()
    }

    @Test
    fun `read throws when not initialized`() {
        assertThrows<CnosError> {
            Cnos.read("value.server.port")
        }
    }

    @Test
    fun `setDefaultRuntime makes read work`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)

        val port = Cnos.value("server.port")
        assertTrue(port.isPresent)
        assertEquals(3000L, (port.get() as Number).toLong())
    }

    @Test
    fun `defaultRuntime returns the set instance`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)
        assertSame(rt, Cnos.defaultRuntime())
    }

    @Test
    fun `ready is idempotent — second call preserves first runtime`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)
        val first = Cnos.defaultRuntime()

        // ready() on an already-initialized singleton keeps the same instance
        Cnos.ready()
        assertSame(first, Cnos.defaultRuntime())
    }

    @Test
    fun `require throws on missing key`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)
        assertThrows<CnosError> {
            Cnos.require("value.does.not.exist")
        }
    }

    @Test
    fun `readOr returns fallback on missing key`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)
        val result = Cnos.readOr("value.missing", "default-val")
        assertEquals("default-val", result)
    }

    @Test
    fun `registerRuntimeProvider is delegated to runtime`() {
        val withRequest = """
        {
          "version": 1, "workspace": "base", "profile": "local",
          "resolvedAt": "2024-01-01T00:00:00Z", "configHash": "abc123",
          "values": { "app.host": "default.host" },
          "derived": {
            "app.effectiveHost": {
              "expr": "coalesce(request.headers.host, value.app.host)",
              "deps": ["value.app.host"],
              "runtimeRefs": ["request.headers.host"]
            }
          },
          "secretRefs": {}, "publicKeys": [], "runtimeNamespaces": ["request"],
          "meta": { "workspace": "base", "profile": "local", "cnos_version": "1.14.0" }
        }""".trimIndent()

        val rt = CnosRuntime.loadProjection(withRequest.toByteArray())
        Cnos.setDefaultRuntime(rt)

        Cnos.registerRuntimeProvider("request") { path ->
            if (path == "headers.host") "console.kitsy.local" else null
        }

        val host = Cnos.value("app.effectiveHost")
        assertTrue(host.isPresent)
        assertEquals("console.kitsy.local", host.get())
    }

    @Test
    fun `format delegates to runtime`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)

        val msg = Cnos.format("App: \${value.app.name}")
        assertEquals("App: cnos-kotlin", msg)
    }

    @Test
    fun `toPublicEnv includes promoted keys`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)

        val env = Cnos.toPublicEnv(ToPublicEnvOptions(framework = "vite"))
        assertEquals("cnos-kotlin", env["VITE_APP_NAME"])
    }

    @Test
    fun `resetDefaultRuntime clears the instance`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)
        Cnos.resetDefaultRuntime()
        assertThrows<CnosError> { Cnos.read("value.app.name") }
    }

    @Test
    fun `toObject returns flat key-value map`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)
        val obj = Cnos.toObject()
        assertNotNull(obj)
        assertTrue(obj.containsKey("value.server.port") || obj.containsKey("value"))
    }

    @Test
    fun `refreshSecrets completes without error when no secrets`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)
        assertDoesNotThrow { Cnos.refreshSecrets() }
    }

    @Test
    fun `refreshSecret on missing key is a no-op`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)
        assertDoesNotThrow { Cnos.refreshSecret("does.not.exist") }
    }

    // ================================================================
    // Composition model
    // ================================================================

    // Simulates: root → libA → libB where each library reads from the
    // module-level singleton and only the root calls setDefaultRuntime.

    private fun libB_read(): Optional<Any> = Cnos.value("server.port")
    private fun libA_read(): Optional<Any> = libB_read()

    @Test
    fun `composition root — libraries succeed after root initializes singleton`() {
        val rt = CnosRuntime.loadProjection(MINIMAL.toByteArray())
        Cnos.setDefaultRuntime(rt)    // root initializes once

        val port = libA_read()        // libA → libB → Cnos.value(…)
        assertTrue(port.isPresent)
        assertEquals(3000L, (port.get() as Number).toLong())
    }

    @Test
    fun `composition root — libraries throw before root initializes`() {
        // root has NOT initialized the singleton
        assertThrows<CnosError> { libA_read() }
    }
}
